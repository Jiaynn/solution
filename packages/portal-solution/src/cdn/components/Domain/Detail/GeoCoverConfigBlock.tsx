/**
 * @file Geo Coverage Config component
 * @description 域名加速覆盖区域的切换。 禁止修改若 1. 该域名未进行 ICP 备案
 * @author hejinxin <hejinxin@hejinxin.com>
 */

import React from 'react'
import { observable, computed } from 'mobx'
import { observer } from 'mobx-react'
import { isEmpty } from 'lodash'
import autobind from 'autobind-decorator'
import { useInjection } from 'qn-fe-core/di'
import Store, { observeInjectable as injectable } from 'qn-fe-core/store'
import { useLocalStore, injectProps } from 'qn-fe-core/local-store'
import { ToasterStore as Toaster } from 'portal-base/common/toaster'
import { ICertInfo } from 'portal-base/certificate'
import { UserInfoStore } from 'portal-base/user/account'
import Spin from 'react-icecream/lib/spin'
import Modal from 'react-icecream/lib/modal'
import Button from 'react-icecream/lib/button'

import { isBucketOversea, shouldForbidModifyGeoCover } from 'cdn/transforms/domain'

import { ModalStore } from 'cdn/stores/modal'

import BucketStore from 'cdn/stores/bucket'

import { isQiniu } from 'cdn/constants/env'
import { OperatingState, DomainType, SourceType } from 'cdn/constants/domain'

import TipIcon from 'cdn/components/TipIcon'
import DomainGeoCoverInput, * as domainGeoCoverInput from 'cdn/components/Domain/Inputs/GeoCoverInput'

import DomainApis, { IDomainDetail } from 'cdn/apis/domain'

import './style.less'

export interface Props {
  domain: IDomainDetail
  hasIcp: boolean
  certInfo?: ICertInfo
  onConfigOk: () => void
}

export default observer(function GeoCoverConfigBlock(props: Props) {
  const store = useLocalStore(GeoCoverConfigStore, props)
  const { onSubmit, ...restModalProps } = store.geoCoverModal.bind()

  const alertTips = [
    isQiniu && '调整加速区域后，流量价格按对应区域计费',
    '海外暂时不支持 IPv6 访问'
  ].filter(Boolean)
  const geoCoverAlert = (
    <div className="domain-detail-alert">
      {alertTips.join('；')}。
    </div>
  )

  const userInfoStore = useInjection(UserInfoStore)
  const shouldForbid = shouldForbidModifyGeoCover(props.domain, userInfoStore, props.certInfo)

  if (store.isTestDomain) {
    return null
  }

  return (
    <>
      <Button
        onClick={store.openModal}
        className="basic-block-btn"
        disabled={!!shouldForbid}
        type="link"
      >
        {
          store.isDomainProcessing ? '处理中' : '修改'
        }
      </Button>
      { shouldForbid && <TipIcon className="basic-block-tip" tip={shouldForbid} /> }
      <Modal
        title="修改覆盖区域"
        className="basic-block-modal"
        okButtonProps={{
          disabled: store.shouldDisableOkButton
        }}
        onOk={onSubmit}
        {...restModalProps}
      >
        <Spin spinning={store.isLoading}>
          {geoCoverAlert}
          <div className="basic-block-content">
            <DomainGeoCoverInput
              hasIcp={props.hasIcp}
              isCNBucket={store.isCNBucket}
              state={store.state}
            />
          </div>
        </Spin>
      </Modal>
    </>
  )
})

@injectable()
class GeoCoverConfigStore extends Store {
  geoCoverModal = new ModalStore()

  constructor(
    @injectProps() protected props: Props,
    private toasterStore: Toaster,
    private bucketStore: BucketStore,
    private domainApis: DomainApis
  ) {
    super()
    Toaster.bindTo(this, this.toasterStore)
  }

  @observable.ref state = domainGeoCoverInput.createState(this.props.domain.geoCover)

  @computed
  get isLoading() {
    return !this.bucketStore.bucketsFetched
  }

  @computed get isCNBucket() {
    const { sourceQiniuBucket, sourceType } = this.props.domain.source
    const bucketInfo = this.bucketStore.getBucket(sourceQiniuBucket)
    return !isEmpty(bucketInfo) && sourceType === SourceType.QiniuBucket && !isBucketOversea(bucketInfo!.zone)
  }

  @autobind
  @Toaster.handle()
  openModal() {
    return this.geoCoverModal.open()
      .then(() => {
        const name = this.props.domain.name

        return this.domainApis.updateGeoCover(name, {
          domains: [name],
          geoCover: this.state.value,
          notifyUser: true
        })
      })
      .then(() => {
        this.props.onConfigOk()
      })
  }

  /* 测试域名不提供域名覆盖区域修改 */
  @computed get isTestDomain() {
    return this.props.domain.type === DomainType.Test
  }

  @computed get isDomainProcessing() {
    return this.props.domain.operatingState === OperatingState.Processing
  }

  @computed get shouldDisableOkButton() {
    return this.state.value === this.props.domain.geoCover
  }

  init() {
    this.addDisposer(this.state.dispose)
    this.addDisposer(this.geoCoverModal.dispose)
  }
}
