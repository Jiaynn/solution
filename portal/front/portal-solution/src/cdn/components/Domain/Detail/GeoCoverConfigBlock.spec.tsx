/**
 * @file cases for GeoCoverConfig component
 * @author hejinxin <hejinxin@qiniu.com>
 */

import React from 'react'

import { RendererUtils as Renderer } from 'test'

import mockDomainDetail from 'cdn/test/domain-detail-mock'

import GeoCoverConfigBlock from './GeoCoverConfigBlock'

const domain = mockDomainDetail()

const noop = () => {}

it('renders correctly', () => {
  const renderer = new Renderer()
  const tree = renderer.createWithAct(
    <GeoCoverConfigBlock domain={domain} hasIcp onConfigOk={noop} />
  ).toJSON()
  expect(tree).toMatchSnapshot()
})

it('renders correctly with no icp record', () => {
  const renderer = new Renderer()
  const tree = renderer.createWithAct(
    <GeoCoverConfigBlock domain={domain} hasIcp={false} onConfigOk={noop} />
  ).toJSON()
  expect(tree).toMatchSnapshot()
})
