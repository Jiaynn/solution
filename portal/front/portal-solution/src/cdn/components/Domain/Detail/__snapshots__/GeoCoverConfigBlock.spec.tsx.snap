// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<button
  className="ant-btn basic-block-btn ant-btn-link"
  disabled={false}
  onClick={[Function]}
  type="button"
>
  <span>
    修改
  </span>
</button>
`;

exports[`renders correctly with no icp record 1`] = `
<button
  className="ant-btn basic-block-btn ant-btn-link"
  disabled={false}
  onClick={[Function]}
  type="button"
>
  <span>
    修改
  </span>
</button>
`;
